src/Ontology/Oep4.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright © 2017-2022 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Oep4.h"
+#include <HexCoding.h>
+#include <cstdio>
+
+namespace TW::Ontology {
+Oep4::Oep4(const Address addr) noexcept
+    : oep4Contract(addr._data.begin(), addr._data.end()) {
+}
+
+Transaction Oep4::readOnlyMethod(std::string method_name, uint32_t nonce) {
+    auto builder = ParamsBuilder();
+    Address addr(oep4Contract);
+    std::vector<boost::any> args{};
+    auto invokeCode = ParamsBuilder::buildOep4InvokeCode(addr, method_name, args);
+
+    return Transaction(0, 0xD1, nonce, 0, 0, "", invokeCode);
+}
+
+Transaction Oep4::name(uint32_t nonce) {
+    return Oep4::readOnlyMethod("name", nonce);
+}
+
+Transaction Oep4::symbol(uint32_t nonce) {
+    return Oep4::readOnlyMethod("symbol", nonce);
+}
+
+Transaction Oep4::decimals(uint32_t nonce) {
+    return Oep4::readOnlyMethod("decimals", nonce);
+}
+
+Transaction Oep4::totalSupply(uint32_t nonce) {
+    return Oep4::readOnlyMethod("totalSupply", nonce);
+}
+
+Transaction Oep4::balanceOf(const Address& user, uint32_t nonce) {
+    auto builder = ParamsBuilder();
+    Address contract(oep4Contract);
+    Data d(std::begin(user._data), std::end(user._data));
```

Comment:
You can use `Data d(user._data);` instead

Reply:
err, can not compile, two type vector has no ctor for array<20, uint8_t>